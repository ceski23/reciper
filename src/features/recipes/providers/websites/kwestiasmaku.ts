import kwestiasmakuIcon from 'assets/images/providers/kwestia_smaku.png'
import { extractMicrodata } from 'features/recipes/providers/generic/microdata'
import type { RecipesProvider } from 'features/recipes/providers/scrapper'
import type { Recipe } from 'features/recipes/types'
import { isDefined } from 'lib/utils'
import { getTextFromNode } from 'lib/utils/dom'
import { getColorFromImage } from 'lib/utils/images'
import { parseValidNumber } from 'lib/utils/numbers'

export const kwestiasmaku: RecipesProvider = {
	name: 'Kwestia Smaku',
	icon: kwestiasmakuIcon,
	matcher: /https:\/\/www\.kwestiasmaku\.com\/.*/,
	scrape: async doc => {
		const partialRecipe = await extractMicrodata(doc)
		const color = partialRecipe?.image ? await getColorFromImage(partialRecipe.image).catch(() => undefined) : undefined
		const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim()

		const tags = Array
			.from(doc.querySelectorAll('#node-przepis-full-group-kategorie a'))
			.map(elem => elem.textContent?.toLowerCase())
			.filter(isDefined)

		const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji')
		const servings = parseValidNumber(servingsElement?.textContent)

		return {
			...partialRecipe,
			ingredients: parseIngredients(doc),
			instructions: parseInstructions(doc),
			color,
			description,
			tags,
			servings,
			gallery: parseImages(doc),
		}
	},
}

const parseIngredients = (doc: Document): Recipe['ingredients'] => {
	const ingredientsRoot = doc.querySelector('.field-name-field-skladniki')
	if (!ingredientsRoot) return []

	// Multiple ingredients groups
	if (ingredientsRoot.querySelector('.wyroznione')) {
		return Array.from(ingredientsRoot.children).reduce<{
			ingredients: Recipe['ingredients']
			group?: string
		}>(({ ingredients, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextIngredients = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				ingredients: ingredients.concat(nextIngredients),
				group: nextGroup,
			}
		}, {
			ingredients: [],
			group: undefined,
		}).ingredients
	}

	// Signle ingredients group
	const recipeIngredients = Array
		.from(ingredientsRoot.querySelectorAll('ul > li'))
		.map(elem => elem.textContent?.trim())
		.filter(isDefined)
		.map(text => ({ text }))

	return recipeIngredients
}

const parseInstructions = (doc: Document): Recipe['instructions'] => {
	const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie')
	if (!instructionsRoot) return []

	// Multiple instruction groups
	if (instructionsRoot.querySelector('.wyroznione')) {
		return Array.from(instructionsRoot.children).reduce<{
			instructions: Recipe['instructions']
			group?: string
		}>(({ instructions, group }, element) => {
			const nextGroup = (element.tagName === 'DIV' ? element.textContent?.trim() : undefined) ?? group
			const nextInstructions = Array
				.from(element.querySelectorAll('li'))
				.map(elem => getTextFromNode(elem).trim())
				.filter(isDefined)
				.map(text => ({ text, group: nextGroup }))

			return {
				instructions: instructions.concat(nextInstructions),
				group: nextGroup,
			}
		}, {
			instructions: [],
			group: undefined,
		}).instructions
	}

	// Signle instructions group
	return Array
		.from(instructionsRoot.querySelectorAll('ul > li'))
		.map(elem => getTextFromNode(elem).trim())
		.filter(isDefined)
		.map(text => ({ text }))
}

const parseImages = (doc: Document) =>
	Array
		.from(doc.querySelectorAll('.view-zdjecia img'))
		.map(elem => elem.getAttribute('src'))
		.filter(isDefined)
