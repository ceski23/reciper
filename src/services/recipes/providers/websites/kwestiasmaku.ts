/* eslint-disable @typescript-eslint/no-loop-func */
/* eslint-disable @typescript-eslint/brace-style */
/* eslint-disable no-restricted-syntax */
import icon from 'assets/providers/kwestia_smaku.png';

import { Recipe, RecipeIngredient, RecipeInstruction } from 'services/recipes';
import type { Provider, RecipeScrapper } from 'services/recipes/providers';
import microdataScrapper from 'services/recipes/providers/microdata';
import { getColorFromImage } from 'services/recipes/providers/utils';

import { getTextFromNode } from 'utils/dom';
import { nonNullable } from 'utils/guards';
import appLogger from 'utils/logger';
import { removeEmpty } from 'utils/objects';

const log = appLogger.extend('scrapper:kwestiasmaku');

export const KwestiaSmakuProvider: Provider = (() => {
  const name = 'Kwestia Smaku';
  const url = 'https://www.kwestiasmaku.com/';

  const parseIngredients = (doc: Document) => {
    const ingredientsRoot = doc.querySelector('.field-name-field-skladniki');
    if (!ingredientsRoot) return [];

    // Multiple ingredients groups
    if (ingredientsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const ingredients: RecipeIngredient[] = [];

      for (const element of ingredientsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeIngredients = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => elem.textContent?.trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          ingredients.push(...recipeIngredients);
        }
      }

      return ingredients;
    }

    // Signle ingredients group
    const ingredientsElements = ingredientsRoot.querySelectorAll('ul > li');
    const recipeIngredients = Array
      .from(ingredientsElements)
      .map((elem) => elem.textContent?.trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeIngredients;
  };

  const parseInstructions = (doc: Document) => {
    const instructionsRoot = doc.querySelector('.field-name-field-przygotowanie');
    if (!instructionsRoot) return [];

    // Multiple instruction groups
    if (instructionsRoot.querySelector('.wyroznione')) {
      let group: string | undefined;
      const instructions: RecipeInstruction[] = [];

      for (const element of instructionsRoot.children) {
        if (element.tagName === 'DIV') {
          group = element.textContent?.trim() ?? undefined;
        } else {
          const recipeInstructions = Array
            .from(element.querySelectorAll('li'))
            .map((elem) => getTextFromNode(elem).trim())
            .filter(nonNullable)
            .map<RecipeInstruction>((text) => ({ text, group }));

          instructions.push(...recipeInstructions);
        }
      }

      return instructions;
    }

    // Signle instructions group
    const instructionsElements = instructionsRoot.querySelectorAll('ul > li');
    const recipeInstructions = Array
      .from(instructionsElements)
      .map((elem) => getTextFromNode(elem).trim())
      .filter(nonNullable)
      .map((text) => ({ text }));

    return recipeInstructions;
  };

  const parseImages = (doc: Document) => {
    log('looking for images');

    const imageElements = doc.querySelectorAll('.view-zdjecia img');
    const images = Array
      .from(imageElements)
      .map((elem) => elem.getAttribute('src'))
      .filter(nonNullable);

    log('found images:', images);

    return images;
  };

  // eslint-disable-next-line @typescript-eslint/no-shadow
  const scrapper: RecipeScrapper = async (doc) => {
    const microdata = await microdataScrapper(doc);
    const description = doc.querySelector('.field-name-field-uwagi-wstepne')?.textContent?.trim() || undefined;

    const tagsElements = doc.querySelectorAll('#node-przepis-full-group-kategorie a');
    const tags = Array
      .from(tagsElements)
      .map((elem) => elem.textContent?.toLowerCase())
      .filter(nonNullable);

    const servingsElement = doc.querySelector('.field-name-field-ilosc-porcji');
    const servings = servingsElement?.textContent
      ? Number.parseInt(servingsElement.textContent, 10)
      : undefined;

    const color = microdata.image ? await getColorFromImage(microdata.image) : undefined;

    const partial: Partial<Recipe> = {
      ingredients: parseIngredients(doc),
      instructions: parseInstructions(doc),
      color,
      description,
      tags,
      servings,
      gallery: parseImages(doc),
    };

    return Object.assign(microdata, removeEmpty(partial));
  };

  return {
    name, url, scrapper, icon,
  };
})();
