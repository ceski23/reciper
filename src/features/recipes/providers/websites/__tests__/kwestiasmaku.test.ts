import { describe, expect, it } from 'vitest'
import { recipe1, recipe2, recipe3, recipe4, recipe5 } from './kwestiasmaku.fixtures'
import { kwestiasmaku } from '../kwestiasmaku'

describe('should parse recipes from kwestiasmaku', async () => {
	it('should parse recipe for Krem Pomidorowy', async () => {
		expect(await kwestiasmaku.scrape(new DOMParser().parseFromString(recipe1, 'text/html'))).toMatchSnapshot()
	})

	it('should parse recipe for Clafoutis', async () => {
		expect(await kwestiasmaku.scrape(new DOMParser().parseFromString(recipe2, 'text/html'))).toMatchSnapshot()
	})

	it('should parse recipe for Wiosenna sałatka ziemniaczana ze szparagami i rzodkiewkami', async () => {
		expect(await kwestiasmaku.scrape(new DOMParser().parseFromString(recipe3, 'text/html'))).toMatchSnapshot()
	})

	it('should parse recipe for Babeczki twarogowe z jabłkami', async () => {
		expect(await kwestiasmaku.scrape(new DOMParser().parseFromString(recipe4, 'text/html'))).toMatchSnapshot()
	})

	it('should parse recipe for Kurczak w sosie bazyliowym z pomidorkami koktajlowymi', async () => {
		expect(await kwestiasmaku.scrape(new DOMParser().parseFromString(recipe5, 'text/html'))).toMatchSnapshot()
	})
})
