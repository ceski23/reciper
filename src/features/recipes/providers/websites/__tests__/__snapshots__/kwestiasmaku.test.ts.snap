// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse recipes from kwestiasmaku > should parse recipe for Babeczki twarogowe z jabłkami 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Filety drobiowe w kremowym sosie bazyliowym z pomidorkami koktajlowymi.",
  "gallery": [
    "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/kurczak-w-sosie-bazyliowym-01.jpg",
  ],
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/kurczak-w-sosie-bazyliowym-00.jpg",
  "ingredients": [
    {
      "text": "650 g filetów z piersi kurczaka",
    },
    {
      "text": "sól, pieprz, 1/3 łyżeczki papryki w proszku",
    },
    {
      "text": "1 łyżki mąki",
    },
    {
      "text": "2 - 3 ząbki czosnku",
    },
    {
      "text": "1 doniczka bazylii",
    },
    {
      "text": "2 łyżki tartego parmezanu",
    },
    {
      "text": "3 łyżki oliwy",
    },
    {
      "text": "200 g pomidorków koktajlowych",
    },
    {
      "text": "80 ml bulionu drobiowego",
    },
    {
      "text": "80 ml śmietanki 30%",
    },
  ],
  "instructions": [
    {
      "text": "Filety umyć, osuszyć, oczyścić z błonek i kostek. Poporcjować mięso na mniejsze kawałki:
                            przekroić wzdłuż na 2 cieńsze filety, odciąć polędwiczki, następnie rozbić tłuczkiem w
                            najgrubszej części na filety podobnej grubości.",
    },
    {
      "text": "Doprawić solą, pieprzem oraz papryką, następnie oprószyć mąką.",
    },
    {
      "text": "W rozdrabniaczu (mini melakserze) zmiksować obrany czosnek, listki bazylii, parmezan i 2
                            łyżki oliwy. Otrzymane pesto doprawić solą i pieprzem.",
    },
    {
      "text": "Rozgrzać dużą patelnię, wlać 1 łyżkę oliwy i stopniowo wkładać filety. Smażyć po ok. 3
                            minuty z każdej strony, aż kurczak ładnie się zrumieni. W razie potrzeby można obsmażać na
                            raty.",
    },
    {
      "text": "Dodać pokrojone pomidorki koktajlowe i rozmieścić je wkoło kurczaka, następnie smażyć przez
                            minutę.",
    },
    {
      "text": "Na patelnię z kurczakiem wlać bulion i zagotować. Gotować przez ok. 1 minutę, następnie
                            przewrócić filety na drugą stronę i powtórzyć gotowanie.",
    },
    {
      "text": "Dodać śmietankę, wymieszać, a następnie zagotować.",
    },
    {
      "text": "Dodać zmiksowane pesto i wymieszać je z sosem, zagotować i odstawić z ognia.",
    },
  ],
  "name": "Kurczak w sosie bazyliowym z pomidorkami koktajlowymi",
  "prepTime": undefined,
  "rating": 4.74,
  "servings": 4,
  "tags": [
    "filety drobiowe",
    "szybki obiad",
    "obiad na niedzielę",
    "kuchnia włoska",
    "dania główne",
    "kurczak",
    "pomidorki koktajlowe",
    "pesto",
  ],
}
`;

exports[`should parse recipes from kwestiasmaku > should parse recipe for Clafoutis 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Clafoutis czyli rodzaj pieczonego naleśnika z dużą ilością owoców.",
  "gallery": [
    "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/clafoutis-01.jpg",
  ],
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/clafoutis-00.jpg",
  "ingredients": [
    {
      "text": "500 g mieszanki owoców: truskawek, malin i borówek amerykańskich",
    },
    {
      "text": "3 jajka + 2 żółtka (temp. pokojowa)",
    },
    {
      "text": "90 g cukru",
    },
    {
      "text": "120 g mąki pszennej",
    },
    {
      "text": "szczypta soli",
    },
    {
      "text": "70 g roztopionego masła (ostudzonego)",
    },
    {
      "text": "300 ml letniego (lekko ciepłego) mleka",
    },
    {
      "text": "1 łyżeczka ekstraktu z wanilii",
    },
    {
      "text": "cukier puder",
    },
    {
      "text": "dla chętnych: lody waniliowe do podania",
    },
  ],
  "instructions": [
    {
      "text": "Piekarnik nagrzać do 175 stopni C. Przygotować naczynie żaroodporne o
                            średnicy ok. 23 cm lub 20 x 30 cm. Posmarować je masłem i wysypać mąką.",
    },
    {
      "text": "Truskawki i borówki opłukać, osuszyć. Truskawki pokroić na połówki jeśli są duże. Wymieszać
                            wszystkie owoce: truskawki, borówki i maliny.",
    },
    {
      "text": "Jajka i żółtka dokładnie ubić z cukrem na jasną i puszystą masę. Następnie dodać mąkę oraz
                            sól i zmiksować do połączenia się składników.",
    },
    {
      "text": "Dodać masło (roztopione i ostudzone) i energicznie zmiksować. Teraz dodawać w małych
                            ilościach mleko (letnie, lekko ciepłe) i ciągle ubijać. Na koniec dodać ekstrakt z wanilii.",
    },
    {
      "text": "Mieszankę wylać do przygotowanej formy i rozłożyć owoce. Piec przez ok. 40 - 45
                              minut. Jeśli pod koniec naleśnik będzie już za bardzo się rumienił, można
                            zmniejszyć temp. do 170 stopni C.",
    },
    {
      "text": "Posypać cukrem pudrem i podawać np. z lodami.",
    },
  ],
  "name": "Clafoutis",
  "prepTime": undefined,
  "rating": 5,
  "servings": 4,
  "tags": [
    "kuchnia francuska",
    "owoce",
    "maliny",
    "truskawki",
    "borówka amerykańska",
  ],
}
`;

exports[`should parse recipes from kwestiasmaku > should parse recipe for Krem Pomidorowy 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Zupa krem z pomidorów. Pyszna z grzankami z mozzarellą oraz pesto bazyliowym.",
  "gallery": [
    "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/krem-pomidorowy-01.jpg",
  ],
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/krem-pomidorowy-00.jpg",
  "ingredients": [
    {
      "group": undefined,
      "text": "2 łyżki oliwy extra vergine",
    },
    {
      "group": undefined,
      "text": "1 cebula",
    },
    {
      "group": undefined,
      "text": "2 ząbki czosnku",
    },
    {
      "group": undefined,
      "text": "1 marchewka",
    },
    {
      "group": undefined,
      "text": "1 łyżeczka oregano",
    },
    {
      "group": undefined,
      "text": "1 łyżeczka słodkiej papryki",
    },
    {
      "group": undefined,
      "text": "szczypta ostrej papryki",
    },
    {
      "group": undefined,
      "text": "szczypta cukru",
    },
    {
      "group": undefined,
      "text": "60 ml białego wina (dla chętnych)",
    },
    {
      "group": undefined,
      "text": "2 świeże pomidory",
    },
    {
      "group": undefined,
      "text": "500 ml bulionu jarzynowego",
    },
    {
      "group": undefined,
      "text": "2 x 400 g (2 puszki) krojonych pomidorów",
    },
    {
      "group": "Grzanki z mozzarellą",
      "text": "1 ciabatta lub bagietka",
    },
    {
      "group": "Grzanki z mozzarellą",
      "text": "200 g tartej mozzarelli",
    },
    {
      "group": "Pesto bazyliowe",
      "text": "1 pęczek bazylii (z jednej "doniczki")",
    },
    {
      "group": "Pesto bazyliowe",
      "text": "1 ząbek czosku",
    },
    {
      "group": "Pesto bazyliowe",
      "text": "10 g sera parmezan lub grana padano",
    },
    {
      "group": "Pesto bazyliowe",
      "text": "3 łyżki orzechów (np. pinii, nerkowców lub dowolnych)",
    },
    {
      "group": "Pesto bazyliowe",
      "text": "70 ml oliwy extra vergine",
    },
  ],
  "instructions": [
    {
      "group": undefined,
      "text": "W garnku podgrzać oliwę, dodać pokrojoną w kosteczkę cebulę, przepołowiony i obrany czosnek
                            oraz obraną i pokrojoną na cienkie plasterki marchewkę. Smażyć co chwilę mieszając przez
                            około 4 - 5 minut.",
    },
    {
      "group": undefined,
      "text": "Dodać oregano, słodką oraz ostrą paprykę, cukier i wymieszać. Wlać wino jeśli go używamy i
                            gotować przez minutę.",
    },
    {
      "group": undefined,
      "text": "Świeże pomidory obrać i pokroić w kostkę, dodać do garnka. Gotować przez 2 minuty, doprawić
                            solą oraz pieprzem.",
    },
    {
      "group": undefined,
      "text": "Wlać gorący bulion i zagotować. Dodać krojone pomidory z puszki i gotować przez ok. 10 - 12
                            minut.",
    },
    {
      "group": undefined,
      "text": "Zmiksować w blenderze kielichowym na krem, sprawdzić jeszcze doprawienie solą oraz pieprzem
                            i podawać z grzankami z mozzarellą oraz pesto.",
    },
    {
      "group": "Grzanki z mozzarellą i pesto",
      "text": "Nagrzać grill w piekarniku. Pieczywo pokroić na kromki, posypać tartą mozzarellą.",
    },
    {
      "group": "Grzanki z mozzarellą i pesto",
      "text": "Przygotować pesto : liście bazylii oderwać i umieścić w mini rozdrabniaczu -
                            melakserze, dodać obrany czosnek, ser, orzechy oraz oliwę. Doprawić solą oraz pieprzem i
                            zmiksować na pesto.",
    },
    {
      "group": "Grzanki z mozzarellą i pesto",
      "text": "Wstawić grzanki do piekarnika i piec aż pieczywo i ser będą zrumienione, przez około 4
                            minuty.",
    },
    {
      "group": "Grzanki z mozzarellą i pesto",
      "text": "Po upieczeniu posmarować pesto i podawać z gorącą zupą krem.",
    },
  ],
  "name": "Krem pomidorowy",
  "prepTime": undefined,
  "rating": 5,
  "servings": 4,
  "tags": [
    "zupy",
    "zupa pomidorowa",
    "zupy krem",
    "pomidory",
    "przepisy fit",
  ],
}
`;

exports[`should parse recipes from kwestiasmaku > should parse recipe for Kurczak w sosie bazyliowym z pomidorkami koktajlowymi 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": undefined,
  "gallery": [
    "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/babeczki-twarogowe-z-jablkami-01.jpg",
  ],
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/babeczki-twarogowe-z-jablkami-00.jpg",
  "ingredients": [
    {
      "text": "250 g twarogu wiejskiego",
    },
    {
      "text": "4 jajka",
    },
    {
      "text": "1 kg jabłek*",
    },
    {
      "text": "150 g cukru (3/4 szklanki)",
    },
    {
      "text": "1 łyżeczka ekstraktu waniliowego",
    },
    {
      "text": "skórka z cytryny lub pomarańczy",
    },
    {
      "text": "60 ml (1/4 szklanki) oleju roślinnego",
    },
    {
      "text": "300 g mąki pszennej",
    },
    {
      "text": "2 łyżeczki proszku do pieczenia",
    },
  ],
  "instructions": [
    {
      "text": "Jajka i twaróg wyjąć wcześniej z lodówki i ogrzać w temp. pokojowej. Twaróg wyłożyć na
                            talerz i rozgnieść praską.",
    },
    {
      "text": "Jabłka obrać, pokroić na ósemki, usunąć gniazda nasienne, następnie pokroić na plasterki.",
    },
    {
      "text": "Piekarnik nagrzać do 200 stopni C. Formę na muffiny wyłoży papilotkami.",
    },
    {
      "text": "Jajka ubijać z cukrem na jasną i puszystą masę, przez około 5 - 7 minut. Dodać wanilię,
                            skórkę z cytryny, twaróg oraz olej roślinny i zmiksować na mniejszych obrotach miksera, do
                            połączenia się składników.",
    },
    {
      "text": "Dodać mąkę oraz proszek do pieczenia i ponownie zmiksować na małych obrotach miksera, tylko
                            do połączenia się składników w jednolite ciasto.",
    },
    {
      "text": "Do masy dodać około 2/3 ilości jabłek i wymieszać łyżką. Wyłożyć do foremek starając się
                            wyłożyć jak najwięcej ciasta z górką. Na wierzch dać resztę jabłek delikatnie wciskając je w
                            ciasto.",
    },
    {
      "text": "Wstawić do nagrzanego piekarnika i piec przez ok. 25 minut (do suchego
                            patyczka). Po upieczeniu można posypać cukrem pudrem i/lub cynamonem.",
    },
  ],
  "name": "Babeczki twarogowe z jabłkami",
  "prepTime": undefined,
  "rating": 4.5,
  "servings": 12,
  "tags": [
    "jabłka",
    "muffiny i babeczki",
    "twaróg",
  ],
}
`;

exports[`should parse recipes from kwestiasmaku > should parse recipe for Wiosenna sałatka ziemniaczana ze szparagami i rzodkiewkami 1`] = `
{
  "calories": undefined,
  "color": "#ffffff",
  "description": "Do sałatki można podać jajko sadzone lub dodać do niej pokrojone, ugotowane na twardo jajko.
                        Inną opcją jest też dodatek sera feta lub sałatkowego.",
  "gallery": [
    "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/wiosenna-salatka-ziemniaczana-03.jpg",
  ],
  "image": "https://www.kwestiasmaku.com/sites/v123.kwestiasmaku.com/files/wiosenna-salatka-ziemniaczana-00.jpg",
  "ingredients": [
    {
      "group": undefined,
      "text": "500 g ziemniaków",
    },
    {
      "group": undefined,
      "text": "1 pęczek zielonych szparagów",
    },
    {
      "group": undefined,
      "text": "pół główki sałaty masłowej",
    },
    {
      "group": undefined,
      "text": "1 pęczek rzodkiewki",
    },
    {
      "group": undefined,
      "text": "2 - 3 młode cebulki",
    },
    {
      "group": undefined,
      "text": "szczypiorek",
    },
    {
      "group": "Sos bazyliowy",
      "text": "listki z 1 doniczki bazylii",
    },
    {
      "group": "Sos bazyliowy",
      "text": "0,5 ząbka czosnku",
    },
    {
      "group": "Sos bazyliowy",
      "text": "4 łyżki oliwy extra",
    },
    {
      "group": "Sos bazyliowy",
      "text": "sól i pieprz",
    },
    {
      "group": "Sos bazyliowy",
      "text": "2 łyżki majonezu",
    },
  ],
  "instructions": [
    {
      "text": "Ziemniaki obrać i ugotować w osolonej wodzie.",
    },
    {
      "text": "Szaparagi umyć, odłamać twarde końce (są łykowate i należy je usunąć). Szparagi cienko obrać
                            obieraczką do połowy długości zaczynając od grubszego końca. Pokroić na kawałki.",
    },
    {
      "text": "Szparagi gotować w lekko osolonej wodzie, przez ok. 3 minuty.",
    },
    {
      "text": "Sałatę opłukać i osuszyć. Rzodkiewki umyć, przyciąć na końcach, pokroić na plasterki.
                            Cebulkę i szczypiorek pokroić.",
    },
    {
      "text": "Ziemniaki i szparagi odcedzić. Ziemniaki pokroić na plasterki.",
    },
    {
      "text": "Przygotować sos bazyliowy : zmiksować wszystkie składniki oprócz majonezu w
                            rozdrabniaczu (mini melakserze), na koniec wymieszać łyżką z majonezem.",
    },
    {
      "text": "W salaterkach układać warstwami: porwane liście sałaty, ziemniaki, szparagi, rzodkiewki,
                            cebulę i szczypiorek. Skrapiać też sosem. Jeśli sałatkę podajemy od razu, można też
                            doprawiać składniki solą oraz pieprzem.",
    },
  ],
  "name": "Wiosenna sałatka ziemniaczana ze szparagami i rzodkiewkami",
  "prepTime": undefined,
  "rating": 4.89,
  "servings": 2,
  "tags": [
    "lunche",
    "sałatki",
    "lekkie kolacje",
    "rzodkiewka",
    "szparagi",
    "ziemniaki",
    "przepisy fit",
    "wegetariańskie",
    "majówka",
    "grill",
  ],
}
`;
